File: libs/java/common/src/main/java/arda/common/model/PageResponse.java
package arda.common.model;

import java.util.List;

public record PageResponse<T>(
    List<T> items,
    PageMetadata metadata
) {
    public static <T> PageResponse<T> of(List<T> items, int page, int size, long totalElements) {
        int totalPages = (size == 0) ? 0 : (int) Math.ceil((double) totalElements / size);
        return new PageResponse<>(
            items,
            new PageMetadata(
                page,
                size,
                totalElements,
                totalPages,
                page < totalPages - 1,
                page > 0
            )
        );
    }

    public record PageMetadata(
        int page,
        int size,
        long totalElements,
        int totalPages,
        boolean hasNext,
        boolean hasPrevious
    ) {}
}

File: arda-be-java/libs/common/src/main/kotlin/arda/common/model/PageResponse.kt
package arda.common.model

data class PageResponse<T>(
    val items: List<T>,
    val metadata: PageMetadata
) {
    companion object {
        fun <T> of(items: List<T>, page: Int, size: Int, totalElements: Long): PageResponse<T> {
            val totalPages = if (size == 0) 0 else Math.ceil(totalElements.toDouble() / size).toInt()
            return PageResponse(
                items = items,
                metadata = PageMetadata(
                    page = page,
                    size = size,
                    totalElements = totalElements,
                    totalPages = totalPages,
                    hasNext = page < totalPages - 1,
                    hasPrevious = page > 0
                )
            )
        }
    }
}

data class PageMetadata(
    val page: Int,
    val size: Int,
    val totalElements: Long,
    val totalPages: Int,
    val hasNext: Boolean,
    val hasPrevious: Boolean
)
